### PR TITLE
Sync: check schema is available or not

### LINKED ISSUE #8137: Undefined index in class.jetpack-sync-functions.php
First reported in a forum thread: https://wordpress.org/support/topic/notice-undefined-index-48/

> Only in the administration area – Intermittent error:
> Notice: Undefined index: scheme in /xxxxxx/plugins/jetpack/sync/class.jetpack-sync-functions.php on line 210

> apache.error.log :
>PHP Notice: Undefined index: scheme in /xxx/plugins/jetpack/sync/class.jetpack-sync-functions.php on line 210\n', referer: https://xxxx/wp-admin/post.php?post=xxxxx&action=edit
> PHP Notice: Undefined index: scheme in /xxx/plugins/jetpack/sync/class.jetpack-sync-functions.php on line 210\n', referer: https://xxx/wp-admin/post-new.php

### PR DESCRIPTION
**Pull request by Umangvaghela**
Fixes #8137

Use empty() to check if it is available.

### CODE REVIEW

File: sync/class.jetpack-sync-functions.php [hunk 1]
```diff
@@ -203,26 +210,26 @@ public static function get_protocol_normalized_url( $callable, $new_value ) {
 		$option_key = self::HTTPS_CHECK_OPTION_PREFIX . $callable;
 
 		$parsed_url = wp_parse_url( $new_value );
-		if ( ! $parsed_url ) {
+		if ( empty ( $parsed_url['scheme'] ) ) {
```
Comment (zinigor) [inline sync/class.jetpack-sync-functions.php]: We can't just `return` here if there's no schema set for the URL. To avoid warnings let's instead check against empty down there where we first try to access the `scheme` key.

File: sync/class.jetpack-sync-functions.php [hunk 2]
```diff
@@ -47,19 +49,23 @@ public static function sanitize_taxonomy( $taxonomy ) {
 		}
 		// Remove any meta_box_cb if they are not the default wp ones.
 		if ( isset( $cloned_taxonomy->meta_box_cb ) &&
-		     ! in_array( $cloned_taxonomy->meta_box_cb, array( 'post_tags_meta_box', 'post_categories_meta_box' ) ) ) {
+		     ! in_array( $cloned_taxonomy->meta_box_cb, array( 'post_tags_meta_box', 'post_categories_meta_box' ) )
```
Comment (jeherve) [inline sync/class.jetpack-sync-functions.php]: If you meant to clarify the `if` statement here, it might be best to move the first check on its own line as well:

```php
if (
    isset( $cloned_taxonomy->meta_box_cb )
    && ! in_array( $cloned_taxonomy->meta_box_cb, array( 'post_tags_meta_box', 'post_categories_meta_box' ) )
) {
```

File: sync/class.jetpack-sync-functions.php [hunk 3]
```diff
@@ -47,19 +49,23 @@ public static function sanitize_taxonomy( $taxonomy ) {
...
+		) {
 			$cloned_taxonomy->meta_box_cb = null;
 		}
 		// Remove update call back
 		if ( isset( $cloned_taxonomy->update_count_callback ) &&
-		     ! is_null( $cloned_taxonomy->update_count_callback ) ) {
+		     ! is_null( $cloned_taxonomy->update_count_callback )
```
Comment (jeherve) [inline sync/class.jetpack-sync-functions.php]: Same as above.

File: sync/class.jetpack-sync-functions.php [hunk 4]
```diff
@@ -47,19 +49,23 @@ public static function sanitize_taxonomy( $taxonomy ) {
...
+		) {
 			$cloned_taxonomy->update_count_callback = null;
 		}
 		// Remove rest_controller_class if it something other then the default.
-		if ( isset( $cloned_taxonomy->rest_controller_class )  &&
-		     'WP_REST_Terms_Controller' !== $cloned_taxonomy->rest_controller_class ) {
+		if ( isset( $cloned_taxonomy->rest_controller_class ) &&
+		     'WP_REST_Terms_Controller' !== $cloned_taxonomy->rest_controller_class
```
Comment (jeherve) [inline sync/class.jetpack-sync-functions.php]: Same as above.

File: sync/class.jetpack-sync-functions.php [hunk 5]
```diff
@@ -203,26 +210,26 @@ public static function get_protocol_normalized_url( $callable, $new_value ) {
...
 
 		$parsed_url = wp_parse_url( $new_value );
-		if ( ! $parsed_url ) {
+		if ( empty ( $parsed_url['scheme'] ) ) {
 			return $new_value;
 		}
 
-		$scheme = $parsed_url['scheme'];
-		$scheme_history = get_option( $option_key, array() );
+		$scheme           = $parsed_url['scheme'];
```
Comment (zinigor) [inline sync/class.jetpack-sync-functions.php]: This is where we need to see if the key exists, and I'd prefer to use `array_key_exists` here instead of `empty`.

File: sync/class.jetpack-sync-functions.php [hunk 6]
```diff
@@ -203,26 +210,26 @@ public static function get_protocol_normalized_url( $callable, $new_value ) {
...
 
-		$forced_scheme =  in_array( 'https', $scheme_history ) ? 'https' : 'http';
+		$forced_scheme = in_array( 'https', $scheme_history ) ? 'https' : 'http';
 
 		return set_url_scheme( $new_value, $forced_scheme );
 	}
 
 	public static function get_raw_url( $option_name ) {
-		$value = null;
+		$value    = null;
```
Comment (zinigor) [inline sync/class.jetpack-sync-functions.php]: Please avoid unnecessary changes like this. It's OK to remove spaces that shouldn't be there in related code, but these edits make the PR harder to review. Thanks!

File: sync/class.jetpack-sync-functions.php (hunk 5 above)
Comment (Umangvaghela) [reply sync/class.jetpack-sync-functions.php]: @zinigor 
If I use 
if ( array_key_exists ( 'scheme' , $parsed_url) ) {
  $scheme = $parsed_url['scheme'];
} else { 
 $scheme = '';
}
This is fine .

File: sync/class.jetpack-sync-functions.php [hunk 7]
```diff
@@ -206,8 +206,11 @@ public static function get_protocol_normalized_url( $callable, $new_value ) {
 		if ( ! $parsed_url ) {
 			return $new_value;
 		}
-
-		$scheme = $parsed_url['scheme'];
+		if( array_key_exists ( 'scheme' , $parsed_url) ) {
```
Comment (zinigor) [inline sync/class.jetpack-sync-functions.php]: Missing space after `if` and after `$parsed_url`, please fix.

Comment (Umangvaghela) [reply sync/class.jetpack-sync-functions.php]: Okay, Sir

**Comment by Umangvaghela** [comment]
@zinigor 
I add the change which is given to you.

**Comment by zinigor** [comment]
This is ready to go! Thank you.
Note: when merging please squash to get rid of the whitespace changes.
